### ISSUE TITLE
Eloquent model timestamps are treated as Carbon\Carbon|null

### ISSUE THREAD

**Issue by magnetas** [nunomaduro/larastan#1580]
- Larastan Version: 2.5.1
- `--level` used: 3
- Pull request with failing test:

### Description

Larastan treats/resolves eloquent model timestamps being of `Carbon\Carbon|null` type instead of `Illuminate\Support\Carbon`.

Prepared a simple case in a fresh Laravel project. Basically my arrow function says it'll return an `Illuminate\Support\Carbon` instance and that's the eloquent casted `created_at` value. Get the following error:

`Anonymous function should return Illuminate\Support\Carbon but returns Carbon\Carbon|null` 

If I check the actual type (i.e. dd($user->created_at)) it in fact is `Illuminate\Support\Carbon`.

Any clue how this can be sorted out? 
Thank you!

### Laravel code where the issue was found

```php
namespace App;

use \App\Models\User;
use \Illuminate\Support\Carbon;

class SomeClass
{
    public function someMethod(): void
    {
        $users = User::get();

        $timestamps = $users->map(fn (User $user): Carbon => $user->created_at);
    }
}
```

**Comment by szepeviktor** [comment]
Hellooo @magnetas! 👋🏻
Did you know Carbon has built-in PHPStan extensions?
```neon
includes:
    - vendor/nesbot/carbon/extension.neon
```

**Comment by szepeviktor** [comment]
..as for this specific case: try returning a non-nullable value.

**Comment by magnetas** [comment]
> Hellooo @magnetas! 👋🏻 Did you know Carbon has built-in PHPStan extensions?

Hi @szepeviktor! First of all thank you for such a prompt reaction!

I wasn't aware of this, but unfortunately it doesn't help with my case - appended it to the includes list and keep getting the same errors.

**Comment by szepeviktor** [comment]
Maybe we have a 🐞 bug?
https://github.com/nunomaduro/larastan/blob/6327eb91b4256615b0193eb62403d91fc6f2eb98/tests/Type/data/model-properties.php#L60

**Comment by magnetas** [comment]
> ...as for this specific case: try returning a non-nullable value.

well...the thing is eloquent model timestamp fields are nullable (i.e. before persisting them they are null. Also DB columns behind them are nullable). Tried hinting my return as ?Carbon (nullable support one) - no luck.

**Comment by magnetas** [comment]
> 



> Maybe we have a 🐞 bug?
> 
> https://github.com/nunomaduro/larastan/blob/6327eb91b4256615b0193eb62403d91fc6f2eb98/tests/Type/data/model-properties.php#L60

Yes, this seems to be related. If this assertion passes it'd mean that created_at is nullable Carbon\Carbon instance. However it isn't since Laravel has the wrapper (I guess mostly for mocking).

**Comment by magnetas** [comment]
Technically `Date::now()` is used to set the timestamp values. 
And we also have this:

https://github.com/nunomaduro/larastan/blob/6327eb91b4256615b0193eb62403d91fc6f2eb98/tests/Type/data/date-extension.php#L24

which passes as well...

**Comment by magnetas** [comment]
> You can debug it in your function: `\PHPStan\dumpType($user->created_at);`

Neat!

This is what I get:

```
 ------ ---------------------------------
  Line   SomeClass.php
 ------ ---------------------------------
  14     Dumped type: Carbon\Carbon|null
 ------ ---------------------------------
```

**Comment by szepeviktor** [comment]
@canvural Could you help? I'm lost.

**Comment by erikgaal** [comment]
@magnetas is right here. The union of `\Illuminate\Support\Carbon` and `\Carbon\Carbon` is simplified to `\Carbon\Carbon`, because they extend each other. When I refactored this code, the union was already there. However, thinking about this now again I'm doubting whether that is the correct behaviour?

https://github.com/nunomaduro/larastan/blob/3e96560fa98c927460e1d0a308a598bc25aaca09/src/Properties/ModelPropertyExtension.php#L151-L153

**Comment by magnetas** [comment]
Just commented these three lines out in my example project, cleared phpstan cache and ran analysis again - all green.

https://github.com/nunomaduro/larastan/blob/11bdd132565d37e808e4ac985a580b492c2d8887/src/Properties/ModelCastHelper.php#L154-L156

Also `dumpType` now returns proper value - `Illuminate\Support\Carbon|null`

Not sure why the union was needed there in the first place. I think what this part of the cast  helper does is it determines what type is returned when casting datetimes/timestamps and the union should not be needed.

However it may be needed when defining writable type (since \Carbon\Carbon or even string can be used to set the value.

https://github.com/nunomaduro/larastan/blob/11bdd132565d37e808e4ac985a580b492c2d8887/src/Properties/ModelPropertyExtension.php#L146

Might be missing some context here too. If that's the case then maybe the property extension should do the checks and decide whether additional union is needed...

**Comment by santigarcor** [comment]
Do we have any update on this one? I'm having the same issue, at level 3 phpstan resolves the type of casted timestamps to `Carbon\Carbon|null`
